src/signature/parser.sml

```diff
@@ -0,0 +1,99 @@
+(* The grammar can be found at
+   https://github.com/JonPRL/sml-red-jonprl/blob/master/doc/signatures.pdf *)
+
+structure SignatureParser =
+struct
+
+  open ParserCombinators
+  open CharParser
+
+  open StringSignature
+
+  infixr 4 << >>
+  infixr 3 &&
+  infix  2 -- ##
+  infix  2 wth suchthat return guard when
+  infixr 1 || <|> ??
+
+  fun inBrackets p = char #"[" >> p << char #"]"
+  fun inParentheses p = char #"(" >> p << char #")"
+  fun inBraces p = char #"{" >> p << char #"}"
+
+  val skipChars = repeatSkip (space <|> char #"\t" <|> char #"\n")
+
+  val opid = repeat1 letter wth String.implode
+  val sortid = repeat1 letter wth String.implode
+  val term = repeat1 letter wth String.implode
+  val metaid = repeat1 letter wth String.implode
+  val symid = repeat1 letter wth String.implode
+
+  val sortlist = separate sortid (char #"," >> space)
+
+  val valence = opt (opt (inBraces sortlist) &&
+    opt (inBrackets sortlist) << char #".") && sortid when
+      (fn (SOME (SOME s1, SOME s2), s3) => SOME (s1, s2, s3)
```

Comment:
Out of curiosity, if every case here returns `SOME`, can we just change this to return the tuple directly instead?